File: packages/indexer/indexer/lints/runtime.ts
import type { IndexLintFinding, ProjectDefinition } from '@use-crux/core/project-index'
import { definitionEvidence, isRecord } from './finding-helpers'
import { indexLintFinding } from './rules'

/** Built-in lint findings for Runtime Engine target authoring hazards. */
export function runtimeLintFindings(
  definitions: readonly ProjectDefinition[],
  options: { readonly runtimeConfigured?: boolean } = {},
): IndexLintFinding[] {
  const targets = runtimeTargetDefinitions(definitions)
  const flows = definitions.filter((definition) => definition.kind === 'flow')
  return [
    ...duplicateTargetNameFindings(targets),
    ...nonLiteralTargetNameFindings(targets),
    ...targetNotExportedFindings(targets),
    ...closureDeferFindings(flows),
    ...missingRuntimeConfigFindings(flows, options.runtimeConfigured),
    ...nondeterministicCodeFindings(flows),
    ...nonSerializablePayloadFindings(flows),
  ]
}

interface RuntimeTargetDefinition {
  readonly definition: ProjectDefinition
  readonly kind: 'flow' | 'task'
  readonly nameLiteral: boolean
  readonly exported: boolean
}

function duplicateTargetNameFindings(targets: readonly RuntimeTargetDefinition[]): IndexLintFinding[] {
  const byName = new Map<string, RuntimeTargetDefinition[]>()
  for (const target of targets) {
    const list = byName.get(target.definition.name) ?? []
    list.push(target)
    byName.set(target.definition.name, list)
  }

  return [...byName.entries()]
    .filter(([, items]) => items.length > 1)
    .map(([name, items]) => {
      const definitions = items.map((item) => item.definition)
      const primary = definitions[0]
      return indexLintFinding({
        ruleId: 'runtime.duplicate_target_name',
        key: name,
        message: `Runtime target name "${name}" is used by ${definitions.length} flow/task declarations. Durable target names must be unique.`,
        ...(primary?.source ? { source: primary.source } : {}),
        primaryDefinitionId: primary?.id,
        relatedDefinitionIds: definitions.map((definition) => definition.id),
        evidence: definitions.map((definition) =>
          definitionEvidence(definition, 'Runtime target shares this durable name'),
        ),
      })
    })
}

function nonLiteralTargetNameFindings(targets: readonly RuntimeTargetDefinition[]): IndexLintFinding[] {
  return targets
    .filter((target) => !target.nameLiteral)
    .map((target) =>
      indexLintFinding({
        ruleId: 'runtime.non_literal_target_name',
        key: target.definition.id,
        message: `${runtimeTargetLabel(target)} does not use a literal durable target name. Use a literal string so Crux can generate stable runtime artifacts.`,
        ...(target.definition.source ? { source: target.definition.source } : {}),
        primaryDefinitionId: target.definition.id,
        relatedDefinitionIds: [target.definition.id],
        evidence: [definitionEvidence(target.definition, 'Runtime target name is not a literal string')],
      }),
    )
}

function targetNotExportedFindings(targets: readonly RuntimeTargetDefinition[]): IndexLintFinding[] {
  return targets
    .filter((target) => !target.exported)
    .map((target) =>
      indexLintFinding({
        ruleId: 'runtime.target_not_exported',
        key: target.definition.id,
        message: `${runtimeTargetLabel(target)} is not a top-level exported declaration. Generated runtime entries can only import named exports.`,
        ...(target.definition.source ? { source: target.definition.source } : {}),
        primaryDefinitionId: target.definition.id,
        relatedDefinitionIds: [target.definition.id],
        evidence: [definitionEvidence(target.definition, 'Runtime target is not exported')],
      }),
    )
}

function closureDeferFindings(flows: readonly ProjectDefinition[]): IndexLintFinding[] {
  return flows.flatMap((flow) =>
    runtimeUsages(flow)
      .filter((usage) => usage.method === 'defer' && usage.closureTarget === true)
      .map((usage, index) =>
        indexLintFinding({
          ruleId: 'runtime.closure_defer',
          key: `${flow.id}:defer:${index}`,
          message: `Flow "${flow.name}" passes an inline closure to flow.defer(). Durable background work must use an exported runtime task target.`,
          ...(sourceLocation(usage) ? { source: sourceLocation(usage) } : flow.source ? { source: flow.source } : {}),
          primaryDefinitionId: flow.id,
          relatedDefinitionIds: [flow.id],
          evidence: [
            definitionEvidence(flow, 'Flow uses flow.defer with an inline closure target'),
            runtimeUsageEvidence(flow, usage, 'Inline defer closure'),
          ],
        }),
      ),
  )
}

function missingRuntimeConfigFindings(
  flows: readonly ProjectDefinition[],
  runtimeConfigured: boolean | undefined,
): IndexLintFinding[] {
  if (runtimeConfigured !== false) return []
  return flows
    .filter((flow) => runtimeUsages(flow).length > 0)
    .map((flow) =>
      indexLintFinding({
        ruleId: 'runtime.missing_runtime_config',
        key: flow.id,
        message: `Flow "${flow.name}" uses runtime-bound APIs, but this project has no runtime configured.`,
        ...(flow.source ? { source: flow.source } : {}),
        primaryDefinitionId: flow.id,
        relatedDefinitionIds: [flow.id],
        evidence: [
          definitionEvidence(flow, 'Flow uses runtime-bound APIs without runtime config'),
          {
            kind: 'definition',
            label: 'Runtime-bound API calls',
            definitionId: flow.id,
            source: flow.source,
            data: { methods: runtimeUsages(flow).map((usage) => usage.method) },
          },
        ],
      }),
    )
}

function nondeterministicCodeFindings(flows: readonly ProjectDefinition[]): IndexLintFinding[] {
  return flows.flatMap((flow) =>
    nondeterministicCalls(flow).map((call, index) =>
      indexLintFinding({
        ruleId: 'flow.nondeterministic_code',
        key: `${flow.id}:${call.expression}:${index}`,
        message: `Flow "${flow.name}" calls ${call.expression} outside flow.step(). Move nondeterministic reads behind a replayed step.`,
        ...(sourceLocation(call) ? { source: sourceLocation(call) } : flow.source ? { source: flow.source } : {}),
        primaryDefinitionId: flow.id,
        relatedDefinitionIds: [flow.id],
        evidence: [
          definitionEvidence(flow, 'Flow contains nondeterministic code'),
          {
            kind: 'definition',
            label: 'Nondeterministic call',
            definitionId: flow.id,
            source: sourceLocation(call) ?? flow.source,
            data: { expression: call.expression },
          },
        ],
      }),
    ),
  )
}

function nonSerializablePayloadFindings(flows: readonly ProjectDefinition[]): IndexLintFinding[] {
  return flows.flatMap((flow) =>
    runtimeUsages(flow)
      .filter((usage) => typeof usage.nonSerializablePayload === 'string')
      .map((usage, index) =>
        indexLintFinding({
          ruleId: 'runtime.non_serializable_payload',
          key: `${flow.id}:${usage.method}:payload:${index}`,
          message: `Flow "${flow.name}" passes a non-JSON ${usage.nonSerializablePayload} payload to flow.${usage.method}(). Durable payloads must be JSON-serializable.`,
          ...(sourceLocation(usage) ? { source: sourceLocation(usage) } : flow.source ? { source: flow.source } : {}),
          primaryDefinitionId: flow.id,
          relatedDefinitionIds: [flow.id],
          evidence: [
            definitionEvidence(flow, 'Flow passes a non-JSON durable payload'),
            runtimeUsageEvidence(flow, usage, 'Non-serializable payload'),
          ],
        }),
      ),
  )
}

function runtimeTargetDefinitions(definitions: readonly ProjectDefinition[]): RuntimeTargetDefinition[] {
  return definitions.flatMap((definition): RuntimeTargetDefinition[] => {
    if (definition.kind !== 'flow' && definition.kind !== 'task') return []
    const runtimeTarget = definition.metadata?.runtimeTarget
    if (!isRecord(runtimeTarget)) return []
    return [
      {
        definition,
        kind: definition.kind,
        nameLiteral: runtimeTarget.nameLiteral === true,
        exported: runtimeTarget.exported === true,
      },
    ]
  })
}

function runtimeTargetLabel(target: RuntimeTargetDefinition): string {
  return `${target.kind} "${target.definition.name}"`
}

function runtimeUsages(definition: ProjectDefinition): RuntimeUsage[] {
  const value = definition.metadata?.runtimeUsages
  return Array.isArray(value) ? value.filter(isRuntimeUsage) : []
}

function nondeterministicCalls(definition: ProjectDefinition): NondeterministicCall[] {
  const value = definition.metadata?.nondeterministicCalls
  return Array.isArray(value) ? value.filter(isNondeterministicCall) : []
}

interface RuntimeUsage {
  readonly method: 'waitFor' | 'defer' | 'after' | 'untilIdle'
  readonly source?: unknown
  readonly closureTarget?: boolean
  readonly nonSerializablePayload?: string
}

interface NondeterministicCall {
  readonly expression: 'Date.now' | 'Math.random' | 'new Date'
  readonly source?: unknown
}

function isRuntimeUsage(value: unknown): value is RuntimeUsage {
  return (
    isRecord(value) &&
    (value.method === 'waitFor' || value.method === 'defer' || value.method === 'after' || value.method === 'untilIdle')
  )
}

function isNondeterministicCall(value: unknown): value is NondeterministicCall {
  return (
    isRecord(value) &&
    (value.expression === 'Date.now' || value.expression === 'Math.random' || value.expression === 'new Date')
  )
}

function runtimeUsageEvidence(
  definition: ProjectDefinition,
  usage: RuntimeUsage,
  label: string,
): IndexLintFinding['evidence'][number] {
  return {
    kind: 'definition',
    label,
    definitionId: definition.id,
    source: sourceLocation(usage) ?? definition.source,
    data: {
      method: usage.method,
      closureTarget: usage.closureTarget,
      nonSerializablePayload: usage.nonSerializablePayload,
    },
  }
}

function sourceLocation(value: { readonly source?: unknown }): IndexLintFinding['source'] | undefined {
  return isRecord(value.source) &&
    typeof value.source.file === 'string' &&
    typeof value.source.line === 'number' &&
    typeof value.source.column === 'number'
    ? {
        file: value.source.file,
        line: value.source.line,
        column: value.source.column,
      }
    : undefined
}

File: crates/lints/src/rules/runtime.rs
//! Runtime Engine lint rules backed by Project Index metadata.

use std::collections::BTreeMap;

use serde_json::{Map, Value, json};

use crate::builder::{StaticIndexLintBuilder, StaticIndexLintFindingInput, definition_evidence};
use crate::facts::{StaticIndexDefinition, StaticIndexLintFinding};
use crate::helpers::metadata_value;

pub(crate) fn runtime_lint_findings(
    builder: &StaticIndexLintBuilder,
    definitions: &[StaticIndexDefinition],
    runtime_configured: Option<bool>,
) -> Vec<StaticIndexLintFinding> {
    let targets = runtime_targets(definitions);
    let flows = definitions
        .iter()
        .filter(|definition| definition.kind == "flow")
        .collect::<Vec<_>>();
    let mut findings = Vec::new();
    findings.extend(duplicate_target_name_findings(builder, &targets));
    findings.extend(non_literal_target_name_findings(builder, &targets));
    findings.extend(target_not_exported_findings(builder, &targets));
    findings.extend(missing_runtime_config_findings(
        builder,
        &flows,
        runtime_configured,
    ));
    findings.extend(flow_runtime_usage_findings(builder, &flows));
    findings
}

struct RuntimeTarget<'a> {
    definition: &'a StaticIndexDefinition,
    name_literal: bool,
    exported: bool,
}

fn duplicate_target_name_findings(
    builder: &StaticIndexLintBuilder,
    targets: &[RuntimeTarget<'_>],
) -> Vec<StaticIndexLintFinding> {
    let mut by_name = BTreeMap::<&str, Vec<&RuntimeTarget<'_>>>::new();
    for target in targets {
        by_name
            .entry(target.definition.name.as_str())
            .or_default()
            .push(target);
    }
    by_name
        .into_iter()
        .filter(|(_, items)| items.len() > 1)
        .filter_map(|(name, items)| {
            let primary = items[0].definition;
            builder.finding(StaticIndexLintFindingInput {
                rule_id: "runtime.duplicate_target_name",
                key: name,
                message: format!(
                    "Runtime target name \"{}\" is used by {} flow/task declarations. Durable target names must be unique.",
                    name,
                    items.len()
                ),
                source: primary.source.as_ref(),
                primary_definition_id: Some(primary.id.as_str()),
                related_definition_ids: items
                    .iter()
                    .map(|target| target.definition.id.clone())
                    .collect(),
                evidence: items
                    .iter()
                    .map(|target| {
                        definition_evidence(
                            target.definition,
                            "Runtime target shares this durable name",
                        )
                    })
                    .collect(),
                fixes: Vec::new(),
            })
        })
        .collect()
}

fn non_literal_target_name_findings(
    builder: &StaticIndexLintBuilder,
    targets: &[RuntimeTarget<'_>],
) -> Vec<StaticIndexLintFinding> {
    targets
        .iter()
        .filter(|target| !target.name_literal)
        .filter_map(|target| {
            let definition = target.definition;
            builder.finding(StaticIndexLintFindingInput {
                rule_id: "runtime.non_literal_target_name",
                key: definition.id.as_str(),
                message: format!(
                    "{} \"{}\" does not use a literal durable target name. Use a literal string so Crux can generate stable runtime artifacts.",
                    definition.kind, definition.name
                ),
                source: definition.source.as_ref(),
                primary_definition_id: Some(definition.id.as_str()),
                related_definition_ids: vec![definition.id.clone()],
                evidence: vec![definition_evidence(
                    definition,
                    "Runtime target name is not a literal string",
                )],
                fixes: Vec::new(),
            })
        })
        .collect()
}

fn target_not_exported_findings(
    builder: &StaticIndexLintBuilder,
    targets: &[RuntimeTarget<'_>],
) -> Vec<StaticIndexLintFinding> {
    targets
        .iter()
        .filter(|target| !target.exported)
        .filter_map(|target| {
            let definition = target.definition;
            builder.finding(StaticIndexLintFindingInput {
                rule_id: "runtime.target_not_exported",
                key: definition.id.as_str(),
                message: format!(
                    "{} \"{}\" is not a top-level exported declaration. Generated runtime entries can only import named exports.",
                    definition.kind, definition.name
                ),
                source: definition.source.as_ref(),
                primary_definition_id: Some(definition.id.as_str()),
                related_definition_ids: vec![definition.id.clone()],
                evidence: vec![definition_evidence(definition, "Runtime target is not exported")],
                fixes: Vec::new(),
            })
        })
        .collect()
}

fn flow_runtime_usage_findings(
    builder: &StaticIndexLintBuilder,
    flows: &[&StaticIndexDefinition],
) -> Vec<StaticIndexLintFinding> {
    let mut findings = Vec::new();
    for flow in flows {
        for (index, usage) in runtime_usages(flow)
            .into_iter()
            .filter(|usage| usage.method == "defer" && usage.closure_target)
            .enumerate()
        {
            findings.push(flow_usage_finding(
                builder,
                "runtime.closure_defer",
                &format!("{}:defer:{}", flow.id, index),
                flow,
                format!(
                    "Flow \"{}\" passes an inline closure to flow.defer(). Durable background work must use an exported runtime task target.",
                    flow.name
                ),
                "Flow uses flow.defer with an inline closure target",
                "Inline defer closure",
                &usage,
            ));
        }
        for (index, call) in nondeterministic_calls(flow).into_iter().enumerate() {
            if let Some(finding) = builder.finding(StaticIndexLintFindingInput {
                rule_id: "flow.nondeterministic_code",
                key: &format!("{}:{}:{}", flow.id, call.expression, index),
                message: format!(
                    "Flow \"{}\" calls {} outside flow.step(). Move nondeterministic reads behind a replayed step.",
                    flow.name, call.expression
                ),
                source: flow.source.as_ref(),
                primary_definition_id: Some(flow.id.as_str()),
                related_definition_ids: vec![flow.id.clone()],
                evidence: vec![
                    definition_evidence(flow, "Flow contains nondeterministic code"),
                    json!({
                        "kind": "definition",
                        "label": "Nondeterministic call",
                        "definitionId": flow.id,
                        "source": flow.source,
                        "data": { "expression": call.expression },
                    }),
                ],
                fixes: Vec::new(),
            }) {
                findings.push(finding);
            }
        }
        for (index, usage) in runtime_usages(flow)
            .into_iter()
            .filter(|usage| usage.non_serializable_payload.is_some())
            .enumerate()
        {
            findings.push(flow_usage_finding(
                builder,
                "runtime.non_serializable_payload",
                &format!("{}:{}:payload:{}", flow.id, usage.method, index),
                flow,
                format!(
                    "Flow \"{}\" passes a non-JSON {} payload to flow.{}(). Durable payloads must be JSON-serializable.",
                    flow.name,
                    usage.non_serializable_payload.as_deref().unwrap_or("value"),
                    usage.method
                ),
                "Flow passes a non-JSON durable payload",
                "Non-serializable payload",
                &usage,
            ));
        }
    }
    findings
}

fn missing_runtime_config_findings(
    builder: &StaticIndexLintBuilder,
    flows: &[&StaticIndexDefinition],
    runtime_configured: Option<bool>,
) -> Vec<StaticIndexLintFinding> {
    if runtime_configured != Some(false) {
        return Vec::new();
    }
    flows
        .iter()
        .filter_map(|flow| {
            let usages = runtime_usages(flow);
            if usages.is_empty() {
                return None;
            }
            builder.finding(StaticIndexLintFindingInput {
                rule_id: "runtime.missing_runtime_config",
                key: flow.id.as_str(),
                message: format!(
                    "Flow \"{}\" uses runtime-bound APIs, but this project has no runtime configured.",
                    flow.name
                ),
                source: flow.source.as_ref(),
                primary_definition_id: Some(flow.id.as_str()),
                related_definition_ids: vec![flow.id.clone()],
                evidence: vec![
                    definition_evidence(flow, "Flow uses runtime-bound APIs without runtime config"),
                    json!({
                        "kind": "definition",
                        "label": "Runtime-bound API calls",
                        "definitionId": flow.id,
                        "source": flow.source,
                        "data": { "methods": usages.iter().map(|usage| usage.method.as_str()).collect::<Vec<_>>() },
                    }),
                ],
                fixes: Vec::new(),
            })
        })
        .collect()
}

fn flow_usage_finding(
    builder: &StaticIndexLintBuilder,
    rule_id: &str,
    key: &str,
    flow: &StaticIndexDefinition,
    message: String,
    definition_label: &str,
    usage_label: &str,
    usage: &RuntimeUsage,
) -> StaticIndexLintFinding {
    builder
        .finding(StaticIndexLintFindingInput {
            rule_id,
            key,
            message,
            source: flow.source.as_ref(),
            primary_definition_id: Some(flow.id.as_str()),
            related_definition_ids: vec![flow.id.clone()],
            evidence: vec![
                definition_evidence(flow, definition_label),
                json!({
                    "kind": "definition",
                    "label": usage_label,
                    "definitionId": flow.id,
                    "source": flow.source,
                    "data": runtime_usage_data(usage),
                }),
            ],
            fixes: Vec::new(),
        })
        .expect("runtime lint rule descriptor exists")
}

fn runtime_usage_data(usage: &RuntimeUsage) -> Value {
    let mut data = Map::new();
    data.insert("method".to_string(), Value::String(usage.method.clone()));
    if usage.closure_target {
        data.insert("closureTarget".to_string(), Value::Bool(true));
    }
    if let Some(payload) = &usage.non_serializable_payload {
        data.insert(
            "nonSerializablePayload".to_string(),
            Value::String(payload.clone()),
        );
    }
    Value::Object(data)
}

fn runtime_targets(definitions: &[StaticIndexDefinition]) -> Vec<RuntimeTarget<'_>> {
    definitions
        .iter()
        .filter(|definition| definition.kind == "flow" || definition.kind == "task")
        .filter_map(|definition| {
            let target = metadata_value(definition, "runtimeTarget")?;
            Some(RuntimeTarget {
                definition,
                name_literal: target
                    .get("nameLiteral")
                    .and_then(Value::as_bool)
                    .unwrap_or(false),
                exported: target
                    .get("exported")
                    .and_then(Value::as_bool)
                    .unwrap_or(false),
            })
        })
        .collect()
}

struct RuntimeUsage {
    method: String,
    closure_target: bool,
    non_serializable_payload: Option<String>,
}

fn runtime_usages(definition: &StaticIndexDefinition) -> Vec<RuntimeUsage> {
    metadata_value(definition, "runtimeUsages")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|value| {
            Some(RuntimeUsage {
                method: value.get("method")?.as_str()?.to_string(),
                closure_target: value
                    .get("closureTarget")
                    .and_then(Value::as_bool)
                    .unwrap_or(false),
                non_serializable_payload: value
                    .get("nonSerializablePayload")
                    .and_then(Value::as_str)
                    .map(str::to_string),
            })
        })
        .collect()
}

struct NondeterministicCall {
    expression: String,
}

fn nondeterministic_calls(definition: &StaticIndexDefinition) -> Vec<NondeterministicCall> {
    metadata_value(definition, "nondeterministicCalls")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|value| {
            Some(NondeterministicCall {
                expression: value.get("expression")?.as_str()?.to_string(),
            })
        })
        .collect()
}
